src/sendmsg_impl.rs

```diff
@@ -0,0 +1,430 @@
+        self.cmsg_cloexec = v;
+        self
+    }
+
+    /// Do not block (MSG_DONTWAIT)
+    pub fn dont_wait(mut self, v: bool) -> RecvMsgFlags {
+        self.dont_wait = v;
+        self
+    }
+
+    /// Do not remove the retrieved data from the receive queue (the next call will return the same data) (MSG_PEEK)
+    pub fn peek(mut self, v: bool) -> RecvMsgFlags {
+        self.peek = v;
+        self
+    }
+
+    /// Wait for the buffers to be filled (may still be interrupted by a signal or the socket hanging up) (MSG_WAITALL)
+    pub fn wait_all(mut self, v: bool) -> RecvMsgFlags {
+        self.wait_all = v;
+        self
+    }
+
+    fn as_cint(&self) -> libc::c_int {
+        let mut result = 0;
+        if self.cmsg_cloexec { result |= MSG_CMSG_CLOEXEC; }
+        if self.dont_wait { result |= MSG_DONTWAIT; }
+        if self.peek { result |= MSG_PEEK; }
+        if self.wait_all { result |= MSG_WAITALL; }
+        result
+    }
+}
+
+impl RecvMsgResultFlags {
+    /// The returned data marks the end of a record (used for SOCK_SEQPACKET) (MSG_EOR)
+    pub fn end_of_record(&self) -> bool {
+        self.end_of_record
+    }
+
+    /// Some data was discarded due to the provided buffers being too short (MSG_TRUNC)
+    pub fn truncated(&self) -> bool {
+        self.truncated
+    }
+
+    /// Some control data was discarded (MSG_CTRUNC)
+    pub fn control_truncated(&self) -> bool {
+        self.control_truncated
+    }
+
+    fn from_cint(flags: libc::c_int) -> RecvMsgResultFlags {
+        RecvMsgResultFlags {
+            end_of_record: (flags & MSG_EOR) != 0,
+            truncated: (flags & MSG_TRUNC) != 0,
+            control_truncated: (flags & MSG_CTRUNC) != 0,
+        }
+    }
+}
+
+#[repr(C)]
+struct MsgHdr {
+    pub msg_name: *const libc::c_void,
+    pub msg_namelen: libc::socklen_t,
+    pub msg_iov: *mut libc::c_void,
+    pub msg_iovlen: libc::size_t,
+    pub msg_control: *mut libc::c_void,
+    pub msg_controllen: libc::size_t,
+    pub msg_flags: libc::c_int,
+}
+
+#[test]
+fn msghdr_size_correctness() {
+    assert_eq!(raw::msghdr_size as usize, mem::size_of::<MsgHdr>());
+}
+
+#[repr(C)]
+struct IoVec {
+    base: *const libc::c_void,
+    len: libc::size_t,
+}
+
+impl IoVec {
+    fn new(buf: &[u8]) -> IoVec {
+        IoVec {
+            base: buf.as_ptr() as *const libc::c_void,
+            len: buf.len() as libc::size_t,
+        }
+    }
+}
+
+#[test]
+fn iovec_size_correctness() {
+    assert_eq!(raw::iovec_size as usize, mem::size_of::<IoVec>());
+}
+
+#[repr(C)]
+struct CmsgHdr {
+    cmsg_len: libc::size_t,
+    cmsg_level: libc::c_int,
+    cmsg_type: libc::c_int,
+}
+
+#[test]
+fn cmsghdr_size_correctness() {
+    assert_eq!(raw::cmsghdr_size as usize, mem::size_of::<CmsgHdr>());
+}
+
+/// Unix credential that can be sent/received over Unix sockets using `ControlMsg::Credential`
+///
+/// This is a Rust version of `struct ucred` from sys/socket.h
+#[derive(Clone, Debug)]
+pub struct UCred{
```

Comment:
`struct ucred` now exists in `libc`, from version 0.2.3: https://github.com/rust-lang-nursery/libc/commit/4920c7e88ac2dab2adfee2d1f35a3ace492b0daa :)